lib/purescript-ast/package.yaml

```diff
@@ -0,0 +1,61 @@
+name: purescript-ast
+version: '0.13.6'
+synopsis: PureScript Programming Language Abstract Syntax Tree
+description: Defines the underlying syntax of the PureScript Programming Language.
+category: Language
+author: Phil Freeman <paf31@cantab.net>
+maintainer: >
+  Gary Burgess <gary.burgess@gmail.com>,
+  Hardy Jones <jones3.hardy@gmail.com>,
+  Harry Garrood <harry@garrood.me>,
+  Christoph Hegemann <christoph.hegemann1337@gmail.com>,
+  Liam Goodacre <goodacre.liam@gmail.com>,
+  Nathan Faubion <nathan@n-son.com>
+copyright: (c) 2013-17 Phil Freeman, (c) 2014-19 Gary Burgess, (c) other contributors (see CONTRIBUTORS.md)
+license: BSD3
+github: purescript/purescript
+homepage: http://www.purescript.org/
+extra-source-files:
+  - README.md
+dependencies:
+  - aeson >=1.0 && <1.5
+  - base >=4.11 && <4.13
+  - base-compat >=0.6.0
+  - bytestring
+  - containers
+  - deepseq
+  - filepath
+  - microlens-platform >=0.3.9.0 && <0.4
+  - mtl >=2.1.0 && <2.3.0
+  - protolude >=0.1.6 && <0.2.4
+  - scientific >=0.3.4.9 && <0.4
+  - text
+  - vector
+
+library:
+  source-dirs: src
+  ghc-options: -Wall -O2
+  default-extensions:
+    - ConstraintKinds
+    - DataKinds
+    - DeriveFunctor
+    - DeriveFoldable
+    - DeriveTraversable
+    - DeriveGeneric
+    - DerivingStrategies
+    - EmptyDataDecls
+    - FlexibleContexts
+    - KindSignatures
+    - LambdaCase
+    - MultiParamTypeClasses
+    - NoImplicitPrelude
+    - PatternGuards
+    - PatternSynonyms
+    - RankNTypes
+    - RecordWildCards
+    - OverloadedStrings
+    - ScopedTypeVariables
+    - TupleSections
+    - ViewPatterns
```

Comment:
What's our policy on default extensions? I don't think we even use all of these in the AST proper. But then, there's extensions in the modules that are already listed here.

Were I to make a suggestion, it'd be to not use `default-extensions` anymore.
* The tooling around extensions is spotty.
* We cargo cult some extensions like I did here.
* Knowing how a module is going to behave can be restricted to each module.

I don't particularly want to push this change. But it was something I was unclear about and didn't know what to do (so I cargo culted).

Reply:
I would rather stick with default-extensions for almost all extensions other than TemplateHaskell; it can be a bit annoying if things like TupleSections or ViewPatterns are enabled on an ad-hoc basis from file to file and you end up with stylistic differences just based on what extensions happened to be enabled in that file. Also, they're a big source of merge conflicts in my experience.

Reply:
Okay. It sounds like your qualm is with the third bullet. That's understandable. As mentioned, I'm not going to push this change. If I work up the gumption to suggest the change, we can have that discussion in a separate PR.

---

We rewrite extensions that exist in the `package.yaml` file in individual modules and we carry around unused extensions in the `package.yaml`. I'm saying that it's unclear why we do these two things. I'm also asking what we should do about it.

Reply:
Right, so to clarify my position I’d say I would be in favour of removing pragmas from source files and using the same set of default-extensions everywhere. It doesn’t bother me if an extension is enabled and not used - I’d prefer that to having different “versions” of the language in use across different parts of the compiler. For example I want to be able to copy some code from one module to another and have it Just Work as far as possible without having to mess with pragmas. Having said that I’m not terribly fussed about this; the current situation is bearable for me.

Reply:
☝️ that was the rationale for expanding `default-expansions` originally. There was a time were there was a total hodgepodge of extensions enabled variously in different modules.

Reply:
Gotcha. Thanks!

How about this:
1. Move all<sup>1</sup> the `purescript-ast`/`purescript-cst` extensions to their respective `package.yaml` files.
1. Move all<sup>1</sup> `purescript` extensions to the root `package.yaml` file.
1. Make a `default-extensions.yaml` file in the root directory.
1. Use either [YAML includes](https://github.com/sol/hpack#not-repeating-yourself) or [`hpack`s default mechanism](https://github.com/sol/hpack#defaults) to include `default-extensions.yaml` in all of the `package.yaml` files.
1. Write a linter (shell script) that checks for extensions in modules and fails if any extensions are enabled<sup>1</sup>.

We can do step 1 in this PR, and steps 2-5 separately. I'm down to make an issue to track that, if so.

<sup>1</sup>: Not including `TemplateHaskell`.

Reply:
I ended up doing effectively that. I also left out `CPP`, `DeriveAnyClass` and `ImplicitParams` for the reasons explained in [this commit](https://github.com/purescript/purescript/pull/3793/commits/a7da01797f750966f933f7c22ef04ced237e90ee). Lemme know if this works or if there's more to do here.

Reply:
Created #3805 to track this.